### PR TITLE
[FIX] base : prevent double membership to 'portal' and 'user' group that is forbidden in V12

### PR DESCRIPTION
**Pull request by legalsylvain**
In v12, a new constrains prevent user to be member for base.group_portal AND base.group_user.
function ``_check_one_user_type`` defined [here](https://github.com/odoo/odoo/blob/12.0/odoo/addons/base/models/res_users.py#L396).
this patch fixes that problem in case of the database contains users that belong to both groups. 
the group_user is conserved. (as it is the group with the most of features).

CC : @fcayre

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Although this can be considered a bad user configuration, I think it's good to auto-fix it. Please include public group as well, as it is the same selectable.

**Comment by legalsylvain** [comment]
@thanks pedro for your quick review.
indeed, updated to handle the three groups assuming group_public < group_portal < group_user.
Does this seem right to you?

File: odoo/addons/base/migrations/12.0.1.3/pre-migration.py [hunk 1]
```diff
@@ -170,6 +170,43 @@ def fill_ir_attachment_res_model_name(cr):
...
+        group_to_remove = conf[0]
+        group_to_keep = conf[1]
+        openupgrade.logged_query(
+            cr, """
+                DELETE FROM res_groups_users_rel
+                WHERE
+                gid = (
+                    SELECT res_id
+                    FROM ir_model_data
+                    WHERE module = 'base' AND name = '%s'
```
Comment (pedrobaeza) [inline odoo/addons/base/migrations/12.0.1.3/pre-migration.py]: `pyscopg2` places the quotes when needed, so remove them (in all places)

```suggestion
                    WHERE module = 'base' AND name = %s
```

Comment (legalsylvain) [reply odoo/addons/base/migrations/12.0.1.3/pre-migration.py]: you're right. Fixed. thanks !

**Review by pedrobaeza — APPROVED** [review]
Code review

**Comment by legalsylvain** [comment]
thanks. please wait before merging. I'll test the new algorithm in the next upgrade.

**Review by fcayre — APPROVED** [review]
Nice to have this in openupgrade instead of fixing this manually in the original data before migrating, thanks.

File: odoo/addons/base/migrations/12.0.1.3/pre-migration.py [hunk 2]
```diff
@@ -170,6 +170,43 @@ def fill_ir_attachment_res_model_name(cr):
...
+                gid = (
+                    SELECT res_id
+                    FROM ir_model_data
+                    WHERE module = 'base' AND name = %s
+                )
+                AND uid IN (
+                    SELECT uid FROM res_groups_users_rel WHERE gid IN (
+                        SELECT res_id
+                        FROM ir_model_data
+                        WHERE module = 'base'
```
Comment (MiquelRForgeFlow) [inline odoo/addons/base/migrations/12.0.1.3/pre-migration.py:200]:
```suggestion
                        WHERE module = 'base' AND model = 'res.groups'
```

File: odoo/addons/base/migrations/12.0.1.3/pre-migration.py [hunk 3]
```diff
@@ -170,6 +170,43 @@ def fill_ir_attachment_res_model_name(cr):
...
+        group_to_remove = conf[0]
+        group_to_keep = conf[1]
+        openupgrade.logged_query(
+            cr, """
+                DELETE FROM res_groups_users_rel
+                WHERE
+                gid = (
+                    SELECT res_id
+                    FROM ir_model_data
+                    WHERE module = 'base' AND name = %s
```
Comment (MiquelRForgeFlow) [inline odoo/addons/base/migrations/12.0.1.3/pre-migration.py:194]:
```suggestion
                    WHERE module = 'base' AND model = 'res.groups' AND name = %s
```

**Review by MiquelRForgeFlow — APPROVED** [review]

File: odoo/addons/base/migrations/12.0.1.3/pre-migration.py (hunk 2 above)
Comment (legalsylvain) [reply odoo/addons/base/migrations/12.0.1.3/pre-migration.py:200]: I don't get it. full xml id (module + name) is unique by design.

**Pull request merged**